File: src/template_project/analysis/predict_template.py
"""Functions for predicting outcomes based on the estimated model."""

import numpy as np
import pandas as pd
from numpy.typing import NDArray
from statsmodels.discrete.discrete_model import BinaryResultsWrapper


def predict_prob_by_age(
    data: pd.DataFrame,
    model: BinaryResultsWrapper,
    group: str,
) -> pd.DataFrame:
    """Predict smoking probability for varying age values.

    For each group value in column data[group] we create new data that runs through a
    grid of age values from data.age.min() to data.age.max() and fixes all column
    values to the ones returned by data.mode(), except for the group column.

    Args:
        data (pandas.DataFrame): The data set.
        model (statsmodels.base.model.Results): The fitted model.
        group (str): Categorical column in data set. We create predictions for each
            unique value in column data[group]. Cannot be 'age' or 'current_smoker'.

    Returns:
        pandas.DataFrame: Predictions. Has columns 'age' and one column for each
            category in column group.

    """
    age_min: int = data["age"].min()
    age_max: int = data["age"].max()
    age_grid: NDArray[np.int_] = np.arange(age_min, age_max + 1)

    mode = data.mode()

    new_data = pd.DataFrame(age_grid, columns=pd.Index(["age"]))

    cols_to_set = list(set(data.columns) - {group, "age", "current_smoker"})
    new_data = new_data.assign(**dict(mode.loc[0, cols_to_set]))

    result = pd.DataFrame({"age": age_grid})
    for group_value in data[group].unique():
        _new_data = new_data.copy()
        _new_data[group] = group_value
        result[group_value] = model.predict(_new_data)

    return result

File: src/template_project/analysis/predict_template.r
# ======================================================================================
# Functions
# ======================================================================================

predict_prob_by_age <- function(data, model, group) {
  age_min <- min(data[["age"]])
  age_max <- max(data[["age"]])
  age_grid <- seq(age_min, age_max)

  mode <- lapply(data, .mode)
  mode <- data.frame(mode)

  new_data <- data.frame(age = age_grid)
  cols_to_set <- setdiff(colnames(data), c(group, "age", "current_smoker"))
  new_data <- cbind(new_data, mode[cols_to_set])

  predicted <- list(age = age_grid)
  for (group_value in unique(data[[group]])) {
    .new_data <- data.frame(new_data) # create a copy
    .new_data[[group]] <- group_value
    predicted[[group_value]] <- predict(model, .new_data, type = "response")
  }

  predicted <- as.data.frame(predicted)
  return(predicted)
}

.mode <- function(x) {
  unique_values <- unique(x)
  mode <- unique_values[which.max(tabulate(match(x, unique_values)))]
  return(mode)
}

# ======================================================================================
# Get config from pytask
# ======================================================================================

args <- commandArgs(trailingOnly = TRUE)
path_to_yaml <- args[length(args)]
config <- yaml::yaml.load_file(path_to_yaml)

# ======================================================================================
# Main
# ======================================================================================

model <- readRDS(config[["model_path"]])
data <- readRDS(config[["data_path"]])
predicted_prob <- predict_prob_by_age(data, model, config[["group"]])
saveRDS(predicted_prob, config[["produces"]])
